### PR TITLE
Shutdown expunged resources cleanup executor properly, and allow other components to configure/start/stop on error

### LINKED ISSUE #9722: Management server state not Down after management service stop [in main]
Management server state not Down after management service stop [in main], also noticed some components are not stopped during service stop. 

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Manager Components: ClusterManagerImpl, ClusterServiceServletAdapter, etc
~~~

##### CLOUDSTACK VERSION

~~~
4.20.0.0-SNAPSHOT
~~~

##### OS / ENVIRONMENT

MS OL8 + KVM OL8, Advanced Networking.

##### SUMMARY

Management server state not Down after management service stop [in main], also noticed some components are not stopped during service stop. 

Manager Components: ClusterManagerImpl, ClusterServiceServletAdapter, etc are not stopped. Didn't notice the below logs after stop.

```
stopping bean ClusterManagerImpl
stopping bean ClusterServiceServletAdapter
```

From 4.20.0.0-SNAPSHOT (Mgmt2 service stopped) =>

MgmtServers_4 20 0 0-SNAPSHOT

```
sureshanaparti$ cat 4.20.0.0-SNAPSHOT_StoppingBeansCheck.txt | wc -l
    546
```

[4.20.0.0-SNAPSHOT_StoppingBeansCheck.txt](https://github.com/user-attachments/files/17102577/4.20.0.0-SNAPSHOT_StoppingBeansCheck.txt)

From 4.19.1.1 env (Mgmt2 service stopped) =>

MgmtServers_4 19 1 1

```
sureshanaparti$ cat 4.19.1.1_StoppingBeansCheck.txt | wc -l
    636
```

[4.19.1.1_StoppingBeansCheck.txt](https://github.com/user-attachments/files/17102588/4.19.1.1_StoppingBeansCheck.txt)

##### STEPS TO REPRODUCE

- Created env with latest main / 4.20.0.0-SNAPSHOT
- Stop the management service from MS  host (`service cloudstack-management stop`) - Notice that management server state is still Up after service is stopped (check in DB - `SELECT id, name, msid, runid, service_ip, service_port, state, last_update, removed FROM cloud.mshost`).

##### EXPECTED RESULTS

~~~
MS state should be in Down state after management service stopped (and all manager components should be stopped for graceful service stop).
~~~

##### ACTUAL RESULTS

~~~
MS state is in Up state after management service stopped (and few manager components are not stopped).
~~~

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR shutdowns expunged resources cleanup executor when obj is available (when config expunged.resources.purge.enabled is true), allows other components to configure/start/stop on error, and adds some logs in component lifecycle classes.

Noticed this exception with custom logs, the remaining components fails to stop after this exception.
```
WARN  [o.a.c.s.l.CloudStackExtendedLifeCycleStart] (SpringContextShutdownHook:null) (logid:) Error on stopping beans - null
java.lang.NullPointerException
        at org.apache.cloudstack.resource.ResourceCleanupServiceImpl.stop(ResourceCleanupServiceImpl.java:584)
        at org.apache.cloudstack.spring.lifecycle.CloudStackExtendedLifeCycle$2.with(CloudStackExtendedLifeCycle.java:105)
        at org.apache.cloudstack.spring.lifecycle.CloudStackExtendedLifeCycle.with(CloudStackExtendedLifeCycle.java:159)
        at org.apache.cloudstack.spring.lifecycle.CloudStackExtendedLifeCycle.stopBeans(CloudStackExtendedLifeCycle.java:101)
        at org.apache.cloudstack.spring.lifecycle.CloudStackExtendedLifeCycleStart.stop(CloudStackExtendedLifeCycleStart.java:32)
        at org.apache.cloudstack.spring.lifecycle.AbstractSmartLifeCycle.stop(AbstractSmartLifeCycle.java:49)
        at org.springframework.context.support.DefaultLifecycleProcessor.doStop(DefaultLifecycleProcessor.java:234)
        at org.springframework.context.support.DefaultLifecycleProcessor.access$300(DefaultLifecycleProcessor.java:54)
        at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.stop(DefaultLifecycleProcessor.java:373)
        at org.springframework.context.support.DefaultLifecycleProcessor.stopBeans(DefaultLifecycleProcessor.java:206)
        at org.springframework.context.support.DefaultLifecycleProcessor.onClose(DefaultLifecycleProcessor.java:129)
        at org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:1069)
        at org.springframework.context.support.AbstractApplicationContext$1.run(AbstractApplicationContext.java:993)
```

Fixes #9722 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

Mgmt2 service stopped =>

MgmtServers_4 20 0 0-SNAPSHOT_Fixed

### How Has This Been Tested?

Manually tested management server start & stop.

```
2024-09-23 18:20:51,111 INFO  [o.a.c.s.l.CloudStackExtendedLifeCycle] (SpringContextShutdownHook:[]) (logid:) stopping bean ClusterServiceServletAdapter
2024-09-23 18:20:51,111 INFO  [o.a.c.s.l.CloudStackExtendedLifeCycle] (SpringContextShutdownHook:[]) (logid:) stopping bean ClusterManagerImpl

[root@ol8 ~]# cat stopping-beans-check.txt | wc -l
665
```

[stopping-beans-check.txt](https://github.com/user-attachments/files/17102657/stopping-beans-check.txt)

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java [hunk 1]
```diff
@@ -69,7 +69,12 @@ public void startBeans() {
         with(new WithComponentLifeCycle() {
             @Override
             public void with(ComponentLifecycle lifecycle) {
-                lifecycle.start();
+                logger.info("starting bean " + lifecycle.getName());
```
Comment (JoaoJandre) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java:72]:
```suggestion
                logger.info("starting bean {}.", lifecycle.getName());
```

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java [hunk 2]
```diff
@@ -69,7 +69,12 @@ public void startBeans() {
...
+                try {
+                    lifecycle.start();
+                } catch (Exception e) {
+                    logger.error(String.format("Error on starting bean %s - %s", lifecycle.getName(), e.getMessage()), e);
```
Comment (JoaoJandre) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java:76]:
```suggestion
                    logger.error("Error on starting bean {} - {}", lifecycle.getName(), e.getMessage(), e);
```

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java [hunk 3]
```diff
@@ -93,13 +98,21 @@ public void with(ComponentLifecycle lifecycle) {
...
+
         with(new WithComponentLifeCycle() {
             @Override
             public void with(ComponentLifecycle lifecycle) {
                 logger.info("stopping bean " + lifecycle.getName());
-                lifecycle.stop();
+                try {
+                    lifecycle.stop();
+                } catch (Exception e) {
+                    logger.error(String.format("Error on stopping bean %s - %s", lifecycle.getName(), e.getMessage()), e);
```
Comment (JoaoJandre) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java:110]:
```suggestion
                    logger.error("Error on stopping bean {} - {}", lifecycle.getName(), e.getMessage(), e);
```

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java [hunk 4]
```diff
@@ -109,10 +122,13 @@ private void configure() {
             @Override
             public void with(ComponentLifecycle lifecycle) {
                 try {
+                    logger.info("configuring bean " + lifecycle.getName());
```
Comment (JoaoJandre) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java:125]:
```suggestion
                    logger.info("configuring bean {}.", lifecycle.getName());
```

File: framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java [hunk 5]
```diff
@@ -109,10 +122,13 @@ private void configure() {
...
                     lifecycle.configure(lifecycle.getName(), lifecycle.getConfigParams());
                 } catch (ConfigurationException e) {
                     logger.error("Failed to configure " +  lifecycle.getName(), e);
                     throw new CloudRuntimeException(e);
+                } catch (Exception e) {
+                    logger.error(String.format("Error on configuring bean %s - %s", lifecycle.getName(), e.getMessage()), e);
```
Comment (JoaoJandre) [inline framework/spring/lifecycle/src/main/java/org/apache/cloudstack/spring/lifecycle/CloudStackExtendedLifeCycle.java:131]:
```suggestion
                    logger.error("Error on configuring bean {} - {}", lifecycle.getName(), e.getMessage(), e);
```

File: framework/spring/module/src/main/java/org/apache/cloudstack/spring/module/factory/CloudStackSpringContext.java [hunk 6]
```diff
@@ -77,8 +77,10 @@ public void registerShutdownHook() {
         for (String appName : contextMap.keySet()) {
             ApplicationContext contex = contextMap.get(appName);
             if (contex instanceof ConfigurableApplicationContext) {
-                logger.trace("registering shutdown hook for bean "+ appName);
+                logger.trace("Registering shutdown hook for bean "+ appName);
```
Comment (JoaoJandre) [inline framework/spring/module/src/main/java/org/apache/cloudstack/spring/module/factory/CloudStackSpringContext.java:80]:
```suggestion
                logger.trace("Registering shutdown hook for bean {}.", appName);
```

File: framework/spring/module/src/main/java/org/apache/cloudstack/spring/module/factory/CloudStackSpringContext.java [hunk 7]
```diff
@@ -77,8 +77,10 @@ public void registerShutdownHook() {
...
                 ((ConfigurableApplicationContext)contex).registerShutdownHook();
+            } else {
+                logger.warn("Shutdown hook not registered for bean " + appName);
```
Comment (JoaoJandre) [inline framework/spring/module/src/main/java/org/apache/cloudstack/spring/module/factory/CloudStackSpringContext.java:83]:
```suggestion
                logger.warn("Shutdown hook not registered for bean {}", appName);
```

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@sureshanaparti , this looks like a good cleanup. I wonder what it fixes other than just the looks of the code, though. We have two shutdown issues:
1. prolonged time of shutdown
2. no status update for shutdown MSses
does this address voth, @sureshanaparti ? (I can see you showed some evidence for the second)

**Comment by sureshanaparti** [comment]
> @sureshanaparti , this looks like a good cleanup. I wonder what it fixes other than just the looks of the code, though. We have two shutdown issues:
> 
> 1. prolonged time of shutdown
> 2. no status update for shutdown MSses
>    does this address voth, @sureshanaparti ? (I can see you showed some evidence for the second)

@DaanHoogland this updates MS status to Down when service is stopped/shutdown.

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Review by rohityadavcloud — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
Tested. Both the status update and the prolonged shutdown time have been fixed by this.

**Comment by rohityadavcloud** [comment]
cc @JoaoJandre - this has been tested, should we merge this?

**Pull request merged**